File: all/3350_adjacent_increasing_subarrays_detection_II/sol_better_space.c
static inline int max(const int e1, const int e2) {
    return e1 > e2 ? e1 : e2;
}

static inline int min(const int e1, const int e2) {
    return e1 < e2 ? e1 : e2;
}

int maxIncreasingSubarrays(int* nums, int numsSize) {
    int answer = 0;
    int current_strictly_increasing_subarray_length = 1;
    int previous_strictly_increasing_subarray_length = 0;

    for (int i = 1; i < numsSize; ++i) {
        if (nums[i] > nums[i - 1]) {
            ++current_strictly_increasing_subarray_length;
        }
        else {
            previous_strictly_increasing_subarray_length = \
                current_strictly_increasing_subarray_length;
            current_strictly_increasing_subarray_length = 1;
        }

        answer = max(
            answer,
            max(
                /* A whole strictly increasing subarray is cut in two half case. */
                current_strictly_increasing_subarray_length >> 1,
                /* Two adjacent but separated strictly increasing subarray case. */
                min(
                    previous_strictly_increasing_subarray_length,
                    current_strictly_increasing_subarray_length
                )
            )
        );
    }

    return answer;
}

File: all/3350_adjacent_increasing_subarrays_detection_II/sol_better_space.py
from typing import List


class Solution:
    def maxIncreasingSubarrays(self, nums: List[int]) -> int:
        answer = 0
        current_strictly_increasing_subarray_length = 1
        previous_strictly_increasing_subarray_length = 0

        for i in range(1, len(nums)):
            if nums[i] > nums[i - 1]:
                current_strictly_increasing_subarray_length += 1
            else:
                previous_strictly_increasing_subarray_length = (
                    current_strictly_increasing_subarray_length
                )
                current_strictly_increasing_subarray_length = 1

            answer = max(
                answer,
                # A whole strictly increasing is cut in two half case.
                current_strictly_increasing_subarray_length >> 1,
                # Two adjacent but separated strictly increasing subarray case.
                min(
                    current_strictly_increasing_subarray_length,
                    previous_strictly_increasing_subarray_length
                )
            )

        return answer


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        [[2, 5, 7, 8, 9, 2, 3, 4, 3, 1]],
        [[1, 2, 3, 4, 4, 4, 4, 5, 6, 7]],
    ]

    for testcase in testcases_list:
        print(sol.maxIncreasingSubarrays(*testcase))
